samcli/hook_packages/terraform/hooks/prepare/resource_linking.py

```diff
@@ -3,8 +3,62 @@
 e.g. linking layers to functions
 """
 
+from dataclasses import dataclass
+from typing import Any, Dict, List, Optional, Union
 import re
-from typing import List
+
+
+@dataclass
+class ConstantValue:
+    value: Any
+
+
+@dataclass
+class References:
+    value: List[str]
+
+
+Expression = Union[ConstantValue, References]
+
+
+@dataclass
+class ResolvedReference:
+    value: str
+    module_address: str
+
+
+@dataclass
+class TFModule:
+    # full path to the module, including parent modules
+    full_address: Optional[str]
+    parent_module: Optional["TFModule"]
+    variables: Dict[str, Expression]
+    resources: List["TFResource"]
+    child_modules: Dict[str, "TFModule"]
+    outputs: Dict[str, Expression]
+
+    # current module's + all child modules' resources
+    def get_all_resources(self) -> List["TFResource"]:
+        all_resources = self.resources.copy()
+        for _, module in self.child_modules.items():
+            all_resources += module.get_all_resources()
+
+        return all_resources
+
+
+@dataclass
+class TFResource:
+    address: str
+    type: str
+    # the module this resource is defined in
+    module: TFModule
+    attributes: Dict[str, Expression]
```

Comment:
left as `attributes` for now, but now the type includes `Expression`.
If we wanted to stick to the terminology of the plan file / configuration, I think we'd have to switch this to `expressions` and also switch module `variables` to `expressions`. But unsure which one is clearer to be honest.